[feat] HomeView MVI 패턴 적용

nakisara01 opened the pull request:
### 설명
View에 저장된 데이터 변경 로직 및 상태들을 Model, View, Intent로 분리했습니다.

### 관련 이슈

Closes #39 

### 작업 내용

- [ ] HomeModel 개발
- [ ] HomeIntent 개발
- [ ] HomeView 리팩토링
- [ ] SwiftDataStorage 추가

### 참고 내용
playbutton 내 동작은 다른 이슈 파서 작업하겠습니다

nakisara01 committed:
[refactor] Modify ScoreCard State

- ScoreCard에서 ...버튼을 눌렀을 때 나타나는 팝오버 내부 버튼 코드를 정리했습니다.
- minWidth, maxWidth, minHeight를 메인에서 받아서 쓰는 대신 isSmall 플래그로 관리하도록 조정했습니다.
- AddScoreButton에서 0 값인지 계속 판단해야하는 부분이 비효율적이어서 isExpanded를 전달받아 t/f를 판단하도록 변경했습니다.

nakisara01 committed:
[refactor] Rename Text Style

- ScoreCard의 title이 Text/Textfield로 나누어져 있던 부분을 TextField 내에서 isEditable 값을 조정해 변경하도록 수정했습니다.

Co-Authored-By: Huigyun Jeong <102240432+Segyun@users.noreply.github.com>

nakisara01 committed:
Merge branch 'dev' into feat/#39

nakisara01 committed:
[refact] Apply Review Content

- recentScore가 3개만 표시되도록 조정했습니다
- extension 내부에서만 사용되는 상태, 변수를 제거하고 메인뷰에서 처리하도록 조정했습니다.

Co-Authored-By: Huigyun Jeong <102240432+Segyun@users.noreply.github.com>

nakisara01 committed:
[feat] Add HomeModel (#39)

nakisara01 committed:
[#39] feat: add SwiftDataStorage

nakisara01 committed:
[#39] feat: add HomeIntent

nakisara01 committed:
[#39] fix: apply model&intent in view

nakisara01 committed:
Merge branch 'dev' into feat/#39

nakisara01 committed:
[#39] fix: replace to HeaderSection

nakisara01 committed:
[#39] refact: refactor HomeView SubView

ehdrbdndns commented on GMG/Scenes/Home/HomeModel.swift:
```diff
@@ -1,2 +1,111 @@
+                    return $0.updatedAt < $1.updatedAt
+                }
+            }
+        }
```
```suggestion
       
       let comparator: (Score, Score) -> Bool = {
               isLatest 
                     ? ($0.updatedAt, $0.createdAt) > ($1.updatedAt, $1.createdAt)
                     : ($0.createdAt, $0.updatedAt) < ($1.createdAt, $1.updatedAt)
       }
       
       return allScores.sorted(by: comparator)
```

ehdrbdndns replied:
요런 느낌으로 정리할 수 있을듯!

nakisara01 replied:
좋네요! 수정하겠습니다!

ehdrbdndns commented on GMG/Scenes/Home/HomeIntent.swift:
```diff
@@ -1,2 +1,40 @@
+        self.model = model
+    }
+    
+    func loadScores(_ context: ModelContext) {
+        model?.fetchScores(context)
+    }
```
modelContext를 인자로 넘기지 말고, Intent를 처음 생성할 때 modelContext를 넘겨주는 방식은 어떤가요?! 아니면 model을 처음 생성할 때 modelContext를 넘겨줘도 괜찮을 것 같고... 조이드의 생각이 궁금해요

nakisara01 replied:
intent에서 관리하는게 좋을 것 같아요! 단순히 행동 위임 정도로 intent를 생각하다 보니 생각치 못한 부분이었는데, View 자체에서 관리하기엔 View는 화면 표시에만 집중시키느게 좋은 것 같아요. Model은 사실 데이터를 처리하는 로직을 관리하는 부분이라고 생각하고 있는데, modelContext를 model에서 처리하면 제가 생각한 역할과는 약간 거리가 멀다고 생각되네용. 그래서 intent에 넣어주면 View의 생명주기를 함께 타면서 model과 SwiftData의 중간 계층이 되어줘서 intent가 적절할 것 같습니다!

nakisara01 replied:
수정하다보니 문제점이 발생했는데 뷰가 그려질 때 onAppear 이전에는 아직 context가 준비되지 않았을 수 있고, 그래서 그 시점에는 HomeIntent를 만들 수가 없어요. 그래서 그 시점에는 EmptyHomeIntent를 선언해 놨다가 시작 후에 교체하려고 하는데 이 제약을 감수하고 Intent 보유를 유지하는게 좋을까요 아니면 다른 방법이 있을까요?

```swift
MARK: - 비어있는 기본값 HomeIntent

struct EmptyHomeIntent: HomeIntentProtocol {
    func loadScores() {}
    func toggleIsLatest() {}
    func deleteScore(_ score: Score) {}
    func selectScore(_ score: Score?) {}
    func renameScore(_ score: Score, newTitle: String) {}
}

MARK: - 이를 적용한 HomeView

import SwiftData
import SwiftUI

struct HomeView: View {
    @Environment(\.modelContext) private var context
    @Environment(Router.self) private var router: Router

    @State private var model: any HomeModelStateProtocol & HomeModelActionProtocol
    @State private var intent: any HomeIntentProtocol = EmptyHomeIntent()

    init() {
        let model: HomeModel = HomeModel()
        self.model = model
    }

    var body: some View {
        ZStack {
            Color.bg1.ignoresSafeArea()

            ScrollView {
                LazyVStack(spacing: Spacing.xl) {
                    HeaderSection(count: model.songCount)
                    RecentFileSection(model: model, intent: intent)
                    AllFilesSection(model: model, intent: intent)
                }
                .safeAreaPadding()
            }
            .scrollIndicators(.hidden)
            .task {
                intent.loadScores()
            }
        }
        .onAppear {
            intent = HomeIntent(model: model, context: context)
            intent.loadScores()
        }
    }
}


```

Segyun replied:
일단 현재 구조대로 하고 #54 에서 같이 수정하는 것이 좋을 듯합니다!

ehdrbdndns replied:
intent에서 modelContext를 Optional로 유지하고, View에서 onAppear로 modelContext넘겨주는 것도 좋은 것 같아요!

EmptyHomeIntent를 따로 만드는건 불필요해보이긴 합니다! 레몬이 말한 것처럼 #54에서 같이 수정해봐요

ehdrbdndns commented on GMG/Scenes/Home/HomeModel.swift:
```diff
@@ -1,2 +1,111 @@
+        let context = storage.modelContext
+        
+        score.title = newTitle
+        try? context?.save()
```
요거 안해도 modelContext에서 자동으로 저장해주는 것으로 알고있어요

```suggestion
        
```

```
ModelContext의 자동 저장 여부를 결정하는 autosaveEnabled 프로퍼티에서 찾을 수 있었다.

자동 저장이 허용된 경우, 즉 암묵적 쓰기에 대한 정의는 데이터가 삽입(insert)되거나 등록(register)된 Model을 변경하면 ModelContext에서 save()를 호출
```

근데 autosaveEnabled는 우리가 따로 설정하지 않아도 default 값이 true이에요
```
   /// - Parameters:
    ///   - modelType: The model type defining the schema used to create the
    ///     model container.
    ///   - inMemory: Whether the container should store data only in memory.
    ///   - isAutosaveEnabled: `true` if autosave is enabled.
    ///   - isUndoEnabled: use `undoManager` in the environment to manage undo
    ///     operations for the model container.
    ///   - onSetup: A callback that will be invoked when the creation of the
    ///     container has has succeeded or failed.
    @MainActor @preconcurrency public func modelContainer(for modelType: any PersistentModel.Type, inMemory: Bool = false, isAutosaveEnabled: Bool = true, isUndoEnabled: Bool = false, onSetup: @escaping (Result<ModelContainer, any Error>) -> Void = { _ in }) -> some View
```

```
ModelContext를 이용해 업데이트 하지 않고 데이터 자체만 업데이트 해주면 자동으로 ModelContext에 기록됨과 동시에 View도 업데이트 된다.
```

참고자료
https://velog.io/@thinkyside/iOS-%EC%89%BD%EA%B3%A0-%ED%9E%99%ED%95%98%EA%B2%8C-SwiftData-%EC%A0%81%EC%9A%A9%ED%95%B4%EB%B3%B4%EA%B8%B0

ehdrbdndns reviewed (commented):
고생했어요~!

Segyun commented on GMG/Scenes/Home/HomeView.swift:
```diff
@@ -313,64 +143,51 @@ extension HomeView {
-            return palette[(safe - 1) % palette.count]
+            let palette: [Color] = [.blue3, .blue4, .blue5, .blue1, .blue2]
+            return palette[i % palette.count]
         }
```
메서드 대신 프로퍼티로 하는 것도 괜찮을 것 같아요~

```suggestion
        private var backgroundColor: Color {
            let palette: [Color] = [.blue3, .blue4, .blue5, .blue1, .blue2]
            return palette[index % palette.count]
        }
```

Segyun commented on GMG/Scenes/Home/HomeView.swift:
```diff
@@ -413,6 +230,173 @@ extension HomeView {
+                            model.isLatest ? Color.black5 : Color.black3
+                        )
+                        .padding(.trailing, 12)
+                        .onTapGesture { if model.isLatest == false { intent.toggleIsLatest() } }
```
메서드 호출 의미를 명확하게 하는 것이 좋을 것 같습니다!

```suggestion
                        .onTapGesture { if model.isLatest == false { intent.setIsLatest(true) } }
```

Segyun commented on GMG/Scenes/Home/HomeView.swift:
```diff
@@ -413,6 +230,173 @@ extension HomeView {
+                            .foregroundStyle(Color.black4.opacity(0.55))
+                    }
+                    .frame(maxWidth: .infinity, alignment: .leading)
+                    .padding(.top, 48)
```
이렇게도 가능해요!

```suggestion
                    VStack(alignment: .leading, spacing: Spacing.xs) {
                        Text("An experience")
                            .foregroundStyle(Color.white3.opacity(0.6))
                        Text("where humming")
                            .foregroundStyle(Color.black8.opacity(0.4))
                        Text("becomes the")
                            .foregroundStyle(Color.black4.opacity(0.4))
                        Text("start of a song")
                            .foregroundStyle(Color.black4.opacity(0.55))
                    }
                    .font(
                        .custom(
                            Typography.WantedSansStd.Bold,
                            size: 42
                        )
                    )
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.top, 48)
```

Segyun commented on GMG/Scenes/Home/HomeView.swift:
```diff
@@ -424,26 +408,13 @@ extension HomeView {
         let minutes = totalSeconds / 60
         let seconds = totalSeconds % 60
         return String(format: "%02d:%02d", minutes, seconds)
     }
```
`ScoreCard`에서만 사용하는 것 같은데 `ScoreCard` 안에서 프로퍼티로 사용하는 것이 어떨까요?

Segyun reviewed (commented):
수고하셨습니다~ `modelContext` 쪽은 #54 에서 수정해야 될 것 같아서 신경 안쓰셔도 될 것 같습니다!

nakisara01 committed:
[#39] refact: change EmptySection font style

nakisara01 committed:
[#39] refact: modify name of toggleIsLatest

nakisara01 committed:
[#39] refact: move extension fucntions

- extension으로 선언한 function인 dataConverter와 formatDuration이 ScoreCard 내에서만 사용해 내부 프로퍼티로 수정했습니다.

ehdrbdndns reviewed (approved):
수고하셨습니다!
